duckietown/dt-duckiebot-interface
# add imu driver for db21m

## Description

Read IMU data on DB21Ms.

### How to test
1. `git clone -b daffy-db21m-imu-driver https://github.com/duckietown/dt-duckiebot-interface.git`
2. `cd [path/to/]dt-duckiebot-interface`
3. Build: `dts devel build -f --pull -H [HOSTNAME]`
4. Run:
    1. stop `duckiebot-interface` container from portainer
    2. `dts devel run -H [HOSTNAME] -M -s -f -- --privileged -it -v /tmp/argus_socket:/tmp/argus_socket`
5. Check:
    1. Open another terminal, `dts start_gui_tools [HOSTNAME]`
    2. `rostopic echo /[HOSTNAME]/imu_node/imu_data`
    3. Try giving the robot angular velocities and linear accelerations (manually or with `dts duckiebot keyboard_control [HOSTNAME]`) and verify with the `imu_data` messages (the previous step)

Typical msg from imu_data:
```
---                                                                                                                                                                                                        [20/17567]
header:                                                                                                                                                                                                              
  seq: 760                                                                                                                                                                                                           
  stamp:                                                                                                                                                                                                             
    secs: 1618840708                                                                                                                                                                                                 
    nsecs:  93475818                                                                                                                                                                                                 
  frame_id: ""                                                                                                                                                                                                  
orientation:                                                                                                                                                                                                         
  x: 0.0                                                                                                                                                                                                             
  y: 0.0                                                                                                                                                                                                             
  z: 0.0                                                                                                                                                                                                             
  w: 0.0                                                                                                                                                                                                             
orientation_covariance: [-1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0, -1.0]                                                                                                                                       
angular_velocity:                                                                                                                                                                                                    
  x: 0.01638244032320685                                                                                                                                                                                             
  y: 0.009190149449603842                                                                                                                                                                                            
  z: -0.00532762286933556                                                                                                                                                                                            
angular_velocity_covariance: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]                                                                                                                                           
linear_acceleration:                                                                                                                                                                                                 
  x: 0.17002982905982905                                                                                                                                                                                             
  y: -0.19158290598290598                                                                                                                                                                                            
  z: -10.058102564102562
linear_acceleration_covariance: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
---
```

## Type of change

Please mark the relevant options with X:

- [ ] This change is constrained only to one package, any interfaces to nodes from other packages are maintained
- [ ] This change is constrained only to packages in this repository, any interfaces to nodes from other repositories are maintained
- [ ] This change is only a documentation update
- [ ] This change restructures the repository (at level higher than package level)
- [ ] This change requires a change in the Duckiebook (if yes, please describe and, if possible, make a pull request for an edit in the book and link it here)

## Tests

Please describe all tests ran and their outcome. It is highly recommended that you at least do the tests listed by default, even if you only changed the documentation.

- [ ] Made the Duckiebot move with keyboard control
- [ ] Visually checked the published camera stream
- [ ] Verified that the lane following demo works
- [ ] Verified that the indefinite navigation demo works

## Style compliance
Please make sure that all your changes conform to the Duckietown (code style)[link] and code (documentation style)[link] guidelines:

- [ ] I checked that this pull request is compliant with the Duckietown code style
- [ ] I checked that this pull request is compliant with the Duckietown code documentation style

## Next steps
Please describe if you have any questions, comments, or doubts towards the Duckietown software guardians regarding implementation, code style, running tests, documentation, etc.

## FOR THE REVIEWERS
__IMPORTANT:__ DO NOT CHANGE THIS SECTION!

- [ ] Functionality (is the change necessary and is it implemented in the most appropriate way): @liampaull or @afdaniele
- [ ] Integration testing (verify the tests and run additional if necessary): @?
- [ ] Code style compliance: @gibernas
- [ ] Documentation compliance: @aleksandarpetrov

_To the reviewers:_ If you deem necessary, request code edits, more tests, documentation update, or additional reviewing. __Do not review your own code!__

_PR Template, last edit 23 Aug 2019_
